File: dart/test/client_test.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

import 'package:recommendai/src/client.dart';
import 'package:recommendai/src/exceptions.dart';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

RecommendAIClient _makeClient(
  Future<http.Response> Function(http.Request) handler,
) {
  return RecommendAIClient(
    apiKey: 'test-key',
    baseUrl: 'http://localhost:8080',
    httpClient: MockClient(handler),
  );
}

http.Response _json(Map<String, dynamic> body, {int status = 200}) =>
    http.Response(jsonEncode(body), status,
        headers: {'content-type': 'application/json'});

// ---------------------------------------------------------------------------
// ping
// ---------------------------------------------------------------------------

void main() {
  group('ping', () {
    test('returns true when server responds 200', () async {
      final client = _makeClient((_) async => http.Response('', 200));
      expect(await client.ping(), isTrue);
    });

    test('returns false when server responds 503', () async {
      final client =
          _makeClient((_) async => http.Response('', 503));
      expect(await client.ping(), isFalse);
    });
  });

  // ─────────────────────────────────────────────────────────────────────────
  group('recommendations', () {
    test('get() returns a list of recommendations', () async {
      final client = _makeClient((req) async {
        expect(req.url.path, equals('/api/recommendations'));
        expect(req.url.queryParameters['user_id'], equals('user1'));
        return _json({
          'recommendations': [
            {'item_id': 'item1', 'score': 0.9, 'reason': 'test', 'metadata': {}}
          ]
        });
      });
      final recs = await client.recommendations.get('user1', limit: 5);
      expect(recs, hasLength(1));
      expect(recs.first.itemId, equals('item1'));
    });

    test('similar() calls correct endpoint', () async {
      final client = _makeClient((req) async {
        expect(req.url.path, equals('/api/recommendations/similar/item99'));
        expect(req.url.queryParameters['limit'], equals('10'));
        return _json({
          'recommendations': [
            {'item_id': 'item2', 'score': 0.8, 'reason': '', 'metadata': {}}
          ]
        });
      });
      final recs =
          await client.recommendations.similar('item99', limit: 10);
      expect(recs, hasLength(1));
      expect(recs.first.itemId, equals('item2'));
    });

    test('popular() passes category query param', () async {
      final client = _makeClient((req) async {
        expect(req.url.path, equals('/api/recommendations/popular'));
        expect(req.url.queryParameters['category'], equals('books'));
        return _json({
          'recommendations': [
            {'item_id': 'book1', 'score': 0.7, 'reason': '', 'metadata': {}}
          ]
        });
      });
      final recs =
          await client.recommendations.popular(limit: 5, category: 'books');
      expect(recs, hasLength(1));
    });
  });

  // ─────────────────────────────────────────────────────────────────────────
  group('items', () {
    test('upsert() posts to /api/items/bulk', () async {
      final client = _makeClient((req) async {
        expect(req.method, equals('POST'));
        expect(req.url.path, equals('/api/items/bulk'));
        return _json({
          'items': [
            {'item_id': 'itemA', 'properties': {}, 'created_at': null, 'updated_at': null}
          ]
        });
      });
      final items = await client.items.upsert([
        {'item_id': 'itemA', 'properties': {'name': 'Book A'}}
      ]);
      expect(items, hasLength(1));
      expect(items.first.itemId, equals('itemA'));
    });
  });

  // ─────────────────────────────────────────────────────────────────────────
  group('error handling', () {
    test('401 response raises AuthenticationException', () async {
      final client = _makeClient((_) async =>
          _json({'detail': 'invalid api key'}, status: 401));
      expect(
        () => client.recommendations.get('u', limit: 5),
        throwsA(isA<AuthenticationException>()),
      );
    });

    test('404 response raises NotFoundException', () async {
      final client = _makeClient(
          (_) async => _json({'detail': 'not found'}, status: 404));
      expect(
        () => client.recommendations.get('u', limit: 5),
        throwsA(isA<NotFoundException>()),
      );
    });

    test('429 response raises RateLimitException', () async {
      final client = _makeClient(
          (_) async => _json({'detail': 'rate limit'}, status: 429));
      expect(
        () => client.recommendations.get('u', limit: 5),
        throwsA(isA<RateLimitException>()),
      );
    });
  });
}

File: go/client_test.go
package recommendai_test

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	recommendai "github.com/recsys-ai/recommendai-go"
)

func newTestServer(t *testing.T, handler http.HandlerFunc) (*httptest.Server, *recommendai.Client) {
	t.Helper()
	srv := httptest.NewServer(handler)
	t.Cleanup(srv.Close)
	c := recommendai.New("test-key", recommendai.WithBaseURL(srv.URL))
	return srv, c
}

func TestPing_OK(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/health" {
			t.Errorf("unexpected path %s", r.URL.Path)
		}
		w.WriteHeader(http.StatusOK)
	})
	if !c.Ping(context.Background()) {
		t.Error("Ping() returned false; expected true")
	}
}

func TestPing_Down(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusServiceUnavailable)
	})
	if c.Ping(context.Background()) {
		t.Error("Ping() returned true; expected false")
	}
}

func TestRecommendations_Get(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/recommendations" {
			t.Errorf("unexpected path %s", r.URL.Path)
		}
		if r.URL.Query().Get("user_id") != "user1" {
			t.Errorf("missing user_id param")
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{
			"recommendations": []map[string]any{
				{"item_id": "item1", "score": 0.9, "reason": "test", "metadata": map[string]any{}},
			},
		})
	})
	recs, err := c.Recommendations.Get(context.Background(), "user1", 5)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if len(recs) != 1 || recs[0].ItemID != "item1" {
		t.Errorf("unexpected recs: %+v", recs)
	}
}

func TestRecommendations_Similar(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/recommendations/similar/item99" {
			t.Errorf("unexpected path %s", r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{
			"recommendations": []map[string]any{
				{"item_id": "item2", "score": 0.8, "reason": "", "metadata": map[string]any{}},
			},
		})
	})
	recs, err := c.Recommendations.Similar(context.Background(), "item99", 10)
	if err != nil {
		t.Fatalf("Similar: %v", err)
	}
	if len(recs) != 1 {
		t.Errorf("expected 1 rec, got %d", len(recs))
	}
}

func TestRecommendations_Popular(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/recommendations/popular" {
			t.Errorf("unexpected path %s", r.URL.Path)
		}
		if r.URL.Query().Get("category") != "books" {
			t.Errorf("missing category param")
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{
			"recommendations": []map[string]any{
				{"item_id": "book1", "score": 0.7, "reason": "", "metadata": map[string]any{}},
			},
		})
	})
	recs, err := c.Recommendations.Popular(context.Background(), 5, "books")
	if err != nil {
		t.Fatalf("Popular: %v", err)
	}
	if len(recs) != 1 {
		t.Errorf("expected 1 rec, got %d", len(recs))
	}
}

func TestItems_Upsert(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			t.Errorf("expected POST, got %s", r.Method)
		}
		if r.URL.Path != "/api/items/bulk" {
			t.Errorf("unexpected path %s", r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{
			"items": []map[string]any{
				{"item_id": "itemA", "properties": map[string]any{}, "created_at": nil, "updated_at": nil},
			},
		})
	})
	items, err := c.Items.Upsert(context.Background(), []map[string]any{
		{"item_id": "itemA", "properties": map[string]any{"name": "Book A"}},
	})
	if err != nil {
		t.Fatalf("Upsert: %v", err)
	}
	if len(items) != 1 || items[0].ItemID != "itemA" {
		t.Errorf("unexpected items: %+v", items)
	}
}

func TestClient_AuthError(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusUnauthorized)
		json.NewEncoder(w).Encode(map[string]any{"detail": "invalid api key"})
	})
	_, err := c.Recommendations.Get(context.Background(), "u", 5)
	if err == nil {
		t.Fatal("expected error, got nil")
	}
}
